# **OSM 2019 Markers Data: Exploratory Data Analysis**

## **Datasets**

The dataset we will use to carry out this study is obtained from a .pbf file taken from the Geofabrik database (https://download.geofabrik.de/europe/spain.html#), a website that allows downloading updated geospatial information for different regions from OpenStreetMap. By clicking on the link, in the *raw directory index* section, we can see all the historical data available on the website. Given that the file corresponding to 2019 was updated in January of that year, we will keep the data from the beginning of 2020, as this way we will avoid losing the information of all the tourist spots added throughout 2019. It is important to note that the information for the Canary Islands is stored separately and must be downloaded from another section of the website (http://download.geofabrik.de/africa/canary-islands.html#).

## **Goal**

Count all the tourist establishments in each Spanish province in 2019, right before the pandemic.

## **Useful Links**

- https://cienciadedatos.net/documentos/py40-puntos-interes-openstreetmap-python
- https://wiki.openstreetmap.org/wiki/Map_features#Tourism

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
data_dir_geom = '/content/gdrive/MyDrive/TFM/Geometry/'
data_dir_geofabrik = '/content/gdrive/MyDrive/TFM/Geofabrik/'
data_dir_new = '/content/gdrive/MyDrive/TFM/New/'

In [ ]:
!pip install osmium

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 13.7 MB/s eta 0:00:00


In [ ]:
!pip install pandas fiona shapely pyproj rtree

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 535.2/535.2 kB 5.1 MB/s eta 0:00:00


In [ ]:
!pip install geopandas

In [ ]:
!pip install geopy

In [ ]:
# Import necessary packages
import osmium as osm
import pandas as pd
import geopandas as gpd
import numpy as np
import itertools
from shapely.geometry import Point, LineString, Polygon
from shapely import wkb
from geopy.distance import distance
import plotly.express as px
import plotly.graph_objects as go

## 1. Load and Explore Data

The first thing we are going to do is to extract the information about the tourist establishments contained in the downloaded file. The first step in extracting a given map element from a .pbf file is to identify how it is stored within OSM. Consulting in the OSM Wikipedia the Map_features list, it can be observed that the tourist establishments are identified with the key *tourism* and that they are stored as elements of type *node* and *way*. In addition, within this category we can find different values, which will help us to filter to keep only the points that we are interested in. Therefore, from all the information stored in the .pbf file, it is necessary to extract only the nodes and ways that have the key *tourism* with the values of interest.

To do this, we will use the **Pyosmium** library, which allows reading and extracting information from OSM files through the **SimpleHandler** class.


In [ ]:
class POIHandler(osm.SimpleHandler):
    '''
    Class to extract information from an osm.pbf file. Only elements identified
    as 'node' or 'area' are extracted. In addition, a filtering can be applied to
    select only those that have tags with a certain key and value.

    The position of the areas is obtained by calculating the centroid of the
    polygon formed by their nodes.

    Arguments
    ---------

    custom_filter: dict
        Dictionary with the keys and values that the elements must have to be
        extracted. For example:

        `{'amenity': ['restaurant', 'bar']}` selects only those elements that
        have the key 'amenity' with value 'restaurant' or 'bar'.

        `{'amenity': ['restaurant', 'bar'], 'building': ['car']}` selects only
        those elements that have the key 'amenity' with value 'restaurant' or
        'bar', or those with the key 'building' with value 'hotel'.
    '''

    def __init__(self, custom_filter=None):
        osm.SimpleHandler.__init__(self)
        self.osm_data = []
        self.custom_filter = custom_filter

        if self.custom_filter:
            for key, value in self.custom_filter.items():
                if isinstance(value, str):
                    self.custom_filter[key] = [value]

    def node(self, node):
        if self.custom_filter is None:
            name = node.tags.get('name', '')
            self.tag_inventory(node, 'node', name)
        else:
            if any([node.tags.get(key) in self.custom_filter[key] for key in self.custom_filter.keys()]):
                name = node.tags.get('name', '')
                self.tag_inventory(node, 'node', name)

    def area(self, area):
        if self.custom_filter is None:
            name = area.tags.get('name', '')
            self.tag_inventory(area, 'area', name)
        else:
            if any([area.tags.get(key) in self.custom_filter[key] for key in self.custom_filter.keys()]):
                name = area.tags.get('name', '')
                self.tag_inventory(area, 'area', name)

    def tag_inventory(self, elem, elem_type, name):
        if elem_type == 'node':
            for tag in elem.tags:
                self.osm_data.append([elem_type,
                                       elem.id,
                                       name,
                                       elem.location.lon,
                                       elem.location.lat,
                                       pd.Timestamp(elem.timestamp),
                                       len(elem.tags),
                                       tag.k,
                                       tag.v])
        if elem_type == 'area':
            try:
                # A Polygon is created with the nodes that form the area to
                # calculate its centroid.
                nodes = list(elem.outer_rings())[0]
                polygon = Polygon([(node.lon, node.lat) for node in nodes])
                for tag in elem.tags:
                    self.osm_data.append([elem_type,
                                           elem.id,
                                           name,
                                           polygon.centroid.x,
                                           polygon.centroid.y,
                                           pd.Timestamp(elem.timestamp),
                                           len(elem.tags),
                                           tag.k,
                                           tag.v])
            except:
                pass

### **Spain**

In [ ]:
# OSM markers Spain: extraction of elements identified with 'tourism'= ['dictionary with values of interest']
poi_handler_sp = POIHandler(custom_filter={'tourism':['alpine_hut', 'apartment', 'camp_pitch', 'camp_site', 'caravan_site', 'chalet',
                                          'guest_house', 'hostel', 'hotel', 'motel', 'wilderness_hut']})
poi_handler_sp.apply_file(data_dir_geofabrik + 'Spain/2019/spain-200101.osm.pbf')

As a result of the extraction, a list is obtained in which each element is in turn a list with the information associated with a node or area. To facilitate the management of the extracted data, we are going to store this information in a pandas dataframe.

In [ ]:
colnames = ['type', 'id', 'name', 'lon', 'lat', 'timestamp','n_tags', 'tag_key',
            'tag_value']
df_poi_sp = pd.DataFrame(poi_handler_sp.osm_data, columns=colnames)
df_poi_sp.head()

type        id name      lon        lat                 timestamp  n_tags  \
0  node  21947483      -3.70235  40.411489 2018-07-15 13:45:36+00:00       5   
1  node  21947483      -3.70235  40.411489 2018-07-15 13:45:36+00:00       5   
2  node  21947483      -3.70235  40.411489 2018-07-15 13:45:36+00:00       5   
3  node  21947483      -3.70235  40.411489 2018-07-15 13:45:36+00:00       5   
4  node  21947483      -3.70235  40.411489 2018-07-15 13:45:36+00:00       5   

            tag_key           tag_value  
0            source          yahoo_maps  
1           tourism           apartment  
2       addr:street  Calle del Calvario  
3     addr:postcode               28012  
4  addr:housenumber                  15

Each row contains an attribute (*tag_key* and *tag_value*) associated to a node. OpenStreetMap has a free tag system so a same node can have an unlimited number of associated attributes. For example, we see that the node with id 21947483 has 5 associated tags.

In [ ]:
df_poi_sp.shape

(112256, 9)

Therefore, to avoid counting a node as many times as tags it has associated, we will be left with only one tag: **tourism**.

In [ ]:
tourism_sp = df_poi_sp[df_poi_sp['tag_key'].isin(['tourism'])]
tourism_sp

type          id                    name       lon        lat  \
1       node    21947483                         -3.702350  40.411489   
6       node    25913327  NH Ciudad de la Imagen -3.788216  40.398389   
8       node    26860857                   Alpha -3.688983  40.317295   
11      node    26860858      Camping l'Alqueria -0.163515  38.985976   
13      node    26860866               Arco Iris -3.907500  40.381111   
...      ...         ...                     ...       ...        ...   
112231  area  1518890438      Hotel Rural Robles -5.662785  40.133844   
112242  area  1518890482     Hotel Ruta Imperial -5.666122  40.127340   
112248  area  1518890772   Cabañas Campus Zuasti -1.747392  42.854190   
112250  area  1518937062        Parador de Verín -7.446803  41.943662   
112253  area  1519296684       La Casa del Risco -4.909609  40.244372   

                       timestamp  n_tags  tag_key    tag_value  
1      2018-07-15 13:45:36+00:00       5  tourism    apartment  
6      2016-10-21 00:50:00+00:00       2  tourism        hotel  
8      2019-03-06 08:15:58+00:00       3  tourism    camp_site  
11     2016-08-28 19:53:29+00:00       2  tourism    camp_site  
13     2012-07-15 17:30:43+00:00       2  tourism    camp_site  
...                          ...     ...      ...          ...  
112231 2019-12-29 21:43:53+00:00      10  tourism        hotel  
112242 2019-12-29 21:43:53+00:00       7  tourism        hotel  
112248 2019-12-29 21:46:48+00:00       2  tourism    camp_site  
112250 2019-12-30 00:22:15+00:00       3  tourism        hotel  
112253 2019-12-30 17:49:15+00:00       4  tourism  guest_house  

[20931 rows x 9 columns]

As we can see, the size of the dataset has been reduced considerably.

### **Canary Islands**

We are going to follow the same procedure with the data from the Canary Islands.

In [ ]:
# OSM markers Canary Islands: extraction of elements identified with 'tourism'= ['dictionary with values of interest']
poi_handler_ci = POIHandler(custom_filter={'tourism':['alpine_hut', 'apartment', 'camp_pitch', 'camp_site', 'caravan_site', 'chalet',
                                          'guest_house', 'hostel', 'hotel', 'motel', 'wilderness_hut']})
poi_handler_ci.apply_file(data_dir_geofabrik + 'Canary_Islands/2019/canary-islands-200101.osm.pbf')

In [ ]:
colnames = ['type', 'id', 'name', 'lon', 'lat', 'timestamp','n_tags', 'tag_key',
            'tag_value']
df_poi_ci = pd.DataFrame(poi_handler_ci.osm_data, columns=colnames)
df_poi_ci.head()

type        id                         name        lon        lat  \
0  node  33104770  allsun Esquinzo Beach Hotel -14.308186  28.073667   
1  node  33104770  allsun Esquinzo Beach Hotel -14.308186  28.073667   
2  node  33104770  allsun Esquinzo Beach Hotel -14.308186  28.073667   
3  node  33105081                      Aldiana -14.316887  28.060005   
4  node  33105081                      Aldiana -14.316887  28.060005   

                  timestamp  n_tags   tag_key                    tag_value  
0 2017-07-20 14:09:17+00:00       3      name  allsun Esquinzo Beach Hotel  
1 2017-07-20 14:09:17+00:00       3   tourism                        hotel  
2 2017-07-20 14:09:17+00:00       3  operator                     alltours  
3 2011-12-13 08:50:05+00:00       2      name                      Aldiana  
4 2011-12-13 08:50:05+00:00       2   tourism                        hotel

In [ ]:
df_poi_ci.shape

(9194, 9)

In [ ]:
tourism_ci = df_poi_ci[df_poi_ci['tag_key'].isin(['tourism'])]
tourism_ci

type          id                                            name  \
1     node    33104770                     allsun Esquinzo Beach Hotel   
4     node    33105081                                         Aldiana   
7     node    33105367                  Iberostar Palace Fuerteventura   
17    node    33105714  disused  hotel (ex: Dunas Club Calela del Sol)   
20    node    33106383                        Iberostar Playa Gaviotas   
...    ...         ...                                             ...   
9170  area    20195871                       Hotel Radisson Blu Resort   
9173  area  1486737346                             H10 Atlantic Sunset   
9180  area  1486752956                        Hard Rock Hotel Tenerife   
9187  area  1495855946                               Albergue El Pinar   
9191  area  1516517358                                 Casa Rural Tato   

            lon        lat                 timestamp  n_tags  tag_key  \
1    -14.308186  28.073667 2017-07-20 14:09:17+00:00       3  tourism   
4    -14.316887  28.060005 2011-12-13 08:50:05+00:00       2  tourism   
7    -14.320375  28.056673 2012-02-14 11:35:21+00:00      10  tourism   
17   -14.288661  28.092809 2011-11-16 08:10:54+00:00       3  tourism   
20   -14.319408  28.056813 2018-03-04 10:19:50+00:00      11  tourism   
...         ...        ...                       ...     ...      ...   
9170 -15.690884  27.772012 2019-10-01 20:56:43+00:00       3  tourism   
9173 -16.774649  28.119212 2019-11-13 09:06:22+00:00       6  tourism   
9180 -16.775028  28.121134 2019-12-21 14:00:05+00:00       7  tourism   
9187 -17.938215  28.701704 2019-11-21 22:13:13+00:00       4  tourism   
9191 -16.401412  28.318998 2019-12-24 01:16:02+00:00       5  tourism   

     tag_value  
1        hotel  
4        hotel  
7        hotel  
17       hotel  
20       hotel  
...        ...  
9170     hotel  
9173     hotel  
9180     hotel  
9187    hostel  
9191     hotel  

[2131 rows x 9 columns]

Again, we select a single tag to count each node only once.

## 2. Prepare Data

To facilitate the study, we will now join the datasets so that we have the tourist establishments of both the mainland and the islands in the same dataset.

In [ ]:
# Concat datasets
tourism = pd.concat([tourism_sp, tourism_ci], ignore_index=True)
tourism

type          id                       name        lon        lat  \
0      node    21947483                             -3.702350  40.411489   
1      node    25913327     NH Ciudad de la Imagen  -3.788216  40.398389   
2      node    26860857                      Alpha  -3.688983  40.317295   
3      node    26860858         Camping l'Alqueria  -0.163515  38.985976   
4      node    26860866                  Arco Iris  -3.907500  40.381111   
...     ...         ...                        ...        ...        ...   
23057  area    20195871  Hotel Radisson Blu Resort -15.690884  27.772012   
23058  area  1486737346        H10 Atlantic Sunset -16.774649  28.119212   
23059  area  1486752956   Hard Rock Hotel Tenerife -16.775028  28.121134   
23060  area  1495855946          Albergue El Pinar -17.938215  28.701704   
23061  area  1516517358            Casa Rural Tato -16.401412  28.318998   

                      timestamp  n_tags  tag_key  tag_value  
0     2018-07-15 13:45:36+00:00       5  tourism  apartment  
1     2016-10-21 00:50:00+00:00       2  tourism      hotel  
2     2019-03-06 08:15:58+00:00       3  tourism  camp_site  
3     2016-08-28 19:53:29+00:00       2  tourism  camp_site  
4     2012-07-15 17:30:43+00:00       2  tourism  camp_site  
...                         ...     ...      ...        ...  
23057 2019-10-01 20:56:43+00:00       3  tourism      hotel  
23058 2019-11-13 09:06:22+00:00       6  tourism      hotel  
23059 2019-12-21 14:00:05+00:00       7  tourism      hotel  
23060 2019-11-21 22:13:13+00:00       4  tourism     hostel  
23061 2019-12-24 01:16:02+00:00       5  tourism      hotel  

[23062 rows x 9 columns]

Next, to ensure that we do not have duplicates, we will delete those records that may have the same name, except in the case where there is an empty value.

In [ ]:
# Eliminate establishments that may be duplicated by name except when it is an empty value

# Filter rows where 'name' is not empty and remove duplicates, keeping the first occurrence
tourism_no_empty = tourism[tourism['name'] != ''].drop_duplicates(subset='name', keep='first')

# Filter rows where 'name' is empty
tourism_empty = tourism[tourism['name'] == '']

# Concatenate the DataFrames
tourism_cleaned = pd.concat([tourism_no_empty, tourism_empty])

tourism_cleaned

type          id                    name        lon        lat  \
1      node    25913327  NH Ciudad de la Imagen  -3.788216  40.398389   
2      node    26860857                   Alpha  -3.688983  40.317295   
3      node    26860858      Camping l'Alqueria  -0.163515  38.985976   
4      node    26860866               Arco Iris  -3.907500  40.381111   
5      node    26860889           Càmping Begur   3.200000  41.940278   
...     ...         ...                     ...        ...        ...   
23019  area  1228903226                         -15.705435  27.787274   
23020  area  1248811338                         -16.755689  28.319000   
23022  area  1250033538                         -16.500868  28.355481   
23024  area  1250730576                         -17.975234  27.809832   
23026  area  1290130532                         -18.001146  27.774778   

                      timestamp  n_tags  tag_key     tag_value  
1     2016-10-21 00:50:00+00:00       2  tourism         hotel  
2     2019-03-06 08:15:58+00:00       3  tourism     camp_site  
3     2016-08-28 19:53:29+00:00       2  tourism     camp_site  
4     2012-07-15 17:30:43+00:00       2  tourism     camp_site  
5     2015-06-26 15:18:06+00:00       2  tourism     camp_site  
...                         ...     ...      ...           ...  
23019 2018-08-06 11:00:10+00:00       2  tourism         hotel  
23020 2018-09-10 13:13:12+00:00       1  tourism  caravan_site  
23022 2018-09-12 08:19:45+00:00       1  tourism     camp_site  
23024 2018-09-13 19:37:45+00:00       2  tourism        chalet  
23026 2018-12-25 07:43:43+00:00       7  tourism        chalet  

[21453 rows x 9 columns]

Before starting to count the number of tourist establishments, there is one last step: to obtain the *geometry* column of each of them from the longitude and latitude values.

In [ ]:
# Obtain geometry for each establishment
geometry = []

for i in range(len(tourism_cleaned)):
    geometry.append(Point(tourism_cleaned.iloc[i]['lon'], tourism_cleaned.iloc[i]['lat']))

tourism_cleaned = tourism_cleaned.set_geometry(geometry)
tourism_cleaned

type          id                    name        lon        lat  \
1      node    25913327  NH Ciudad de la Imagen  -3.788216  40.398389   
2      node    26860857                   Alpha  -3.688983  40.317295   
3      node    26860858      Camping l'Alqueria  -0.163515  38.985976   
4      node    26860866               Arco Iris  -3.907500  40.381111   
5      node    26860889           Càmping Begur   3.200000  41.940278   
...     ...         ...                     ...        ...        ...   
23019  area  1228903226                         -15.705435  27.787274   
23020  area  1248811338                         -16.755689  28.319000   
23022  area  1250033538                         -16.500868  28.355481   
23024  area  1250730576                         -17.975234  27.809832   
23026  area  1290130532                         -18.001146  27.774778   

                      timestamp  n_tags  tag_key     tag_value  \
1     2016-10-21 00:50:00+00:00       2  tourism         hotel   
2     2019-03-06 08:15:58+00:00       3  tourism     camp_site   
3     2016-08-28 19:53:29+00:00       2  tourism     camp_site   
4     2012-07-15 17:30:43+00:00       2  tourism     camp_site   
5     2015-06-26 15:18:06+00:00       2  tourism     camp_site   
...                         ...     ...      ...           ...   
23019 2018-08-06 11:00:10+00:00       2  tourism         hotel   
23020 2018-09-10 13:13:12+00:00       1  tourism  caravan_site   
23022 2018-09-12 08:19:45+00:00       1  tourism     camp_site   
23024 2018-09-13 19:37:45+00:00       2  tourism        chalet   
23026 2018-12-25 07:43:43+00:00       7  tourism        chalet   

                         geometry  
1       POINT (-3.78822 40.39839)  
2       POINT (-3.68898 40.31730)  
3       POINT (-0.16352 38.98598)  
4       POINT (-3.90750 40.38111)  
5        POINT (3.20000 41.94028)  
...                           ...  
23019  POINT (-15.70544 27.78727)  
23020  POINT (-16.75569 28.31900)  
23022  POINT (-16.50087 28.35548)  
23024  POINT (-17.97523 27.80983)  
23026  POINT (-18.00115 27.77478)  

[21453 rows x 10 columns]

In [ ]:
# Select columns of interest
tourism_cleaned = tourism_cleaned[['id', 'geometry']]
tourism_cleaned

id                    geometry
1        25913327   POINT (-3.78822 40.39839)
2        26860857   POINT (-3.68898 40.31730)
3        26860858   POINT (-0.16352 38.98598)
4        26860866   POINT (-3.90750 40.38111)
5        26860889    POINT (3.20000 41.94028)
...           ...                         ...
23019  1228903226  POINT (-15.70544 27.78727)
23020  1248811338  POINT (-16.75569 28.31900)
23022  1250033538  POINT (-16.50087 28.35548)
23024  1250730576  POINT (-17.97523 27.80983)
23026  1290130532  POINT (-18.00115 27.77478)

[21453 rows x 2 columns]

Once this is done, we can represent our points on a map to check in which province each of them is located. That is, we will check if the geometry of each point falls within the geometry of a particular province.

## 3. Represent and Count Data

In this section, we are going to count how many tourist establishments there were in each Spanish province before the pandemic. As we did in previous studies, we will access the geometry of Spain (taking into account that the geometry of Santa Cruz de Tenerife and Las Palmas will be added later) and we will represent and count the different tourist spots per province. First, let's plot all the tourist points identified in the previous section on a map of Spain.

In [ ]:
fig = px.scatter_mapbox(
    tourism_cleaned,
    lat=tourism_cleaned.geometry.y,
    lon=tourism_cleaned.geometry.x,
    size=None,
    #color_continuous_scale='Viridis',
    mapbox_style="open-street-map",
    zoom=5,
    center={"lat": 40.0, "lon": -3.5},
    title="Tourist Establishments in Spain"
)

fig.update_traces(marker=dict(size=15))

fig.update_layout(margin={"r":0,"t":30,"l":0,"b":0})
fig.show()

As shown, the points are distributed across the entire country. Due to the large number of tourist establishments, it is necessary to zoom in to distinguish specific areas.

In [ ]:
spanish_provinces = gpd.read_file(data_dir_geom + 'spanish_provinces.geojson')
spanish_provinces

rotulo    population  time_light_sun  \
0              Melilla  1.173039e+06              12   
1                Ceuta  1.884441e+06               5   
2                Cádiz  5.056655e+05              11   
3               Málaga  1.077906e+06              10   
4              Almería  1.849180e+06              12   
5              Granada  1.589546e+06               6   
6              Sevilla  1.696450e+06              10   
7               Huelva  1.109913e+06               8   
8                 Jaén  1.827838e+06              12   
9              Córdoba  1.714056e+06               5   
10  Castelló/Castellón  1.730377e+06              10   
11              Murcia  1.088511e+06               5   
12       Illes Balears  1.198749e+06              11   
13    Alacant/Alicante  1.756815e+06               8   
14            Albacete  1.053021e+06               8   
15             Badajoz  1.333095e+06               5   
16         Ciudad Real  1.509088e+06               7   
17   València/Valencia  6.645286e+05               8   
18            Gipuzkoa  1.675893e+06               8   
19              Toledo  9.819115e+05              12   
20             Cáceres  1.183986e+06               4   
21              Madrid  1.412665e+06              10   
22              Cuenca  6.688378e+05               7   
23               Ávila  7.632510e+05               4   
24         Guadalajara  1.197631e+06               7   
25           Salamanca  1.332640e+06              10   
26              Teruel  1.972688e+06               5   
27             Segovia  6.763886e+05               4   
28           Tarragona  1.130393e+06               4   
29               Soria  1.823542e+06               9   
30              Zamora  1.732504e+06               4   
31          Valladolid  1.178602e+06              11   
32           Barcelona  6.470960e+05               4   
33              Girona  7.799650e+05               9   
34            La Rioja  1.582246e+06              10   
35             Ourense  1.383243e+06               7   
36             Navarra  1.480422e+06               7   
37            Zaragoza  1.233700e+06               9   
38              Lleida  7.431729e+05               6   
39              Huesca  1.908733e+06               6   
40          Pontevedra  5.586279e+05              10   
41              Burgos  1.231123e+06              12   
42                León  1.244082e+06               6   
43           Cantabria  1.735293e+06               5   
44            Asturias  1.962832e+06              11   
45                Lugo  7.876897e+05              10   
46            A Coruña  9.491082e+05              10   
47             Bizkaia  7.869600e+05               5   
48         Araba/Álava  1.752627e+06              12   
49            Palencia  1.132019e+06              12   

                                             geometry  
0   POLYGON ((-2.93446 35.27442, -2.92929 35.27324...  
1   POLYGON ((-5.33273 35.89270, -5.31086 35.88727...  
2   POLYGON ((-5.14234 37.00387, -5.11510 36.98897...  
3   POLYGON ((-4.32759 37.18435, -4.32333 37.18075...  
4   POLYGON ((-2.18494 37.90366, -2.17201 37.88872...  
5   POLYGON ((-2.34163 38.02603, -2.32309 38.01401...  
6   POLYGON ((-5.72320 38.19244, -5.70785 38.18830...  
7   POLYGON ((-6.18021 37.94103, -6.14518 37.92050...  
8   POLYGON ((-2.55129 38.08412, -2.57299 38.07785...  
9   POLYGON ((-5.00801 38.71523, -4.87696 38.68610...  
10  POLYGON ((-0.07991 40.73335, -0.06476 40.72742...  
11  POLYGON ((-1.11971 38.73774, -1.11563 38.71141...  
12  MULTIPOLYGON (((3.19286 39.93983, 3.15416 39.9...  
13  POLYGON ((0.01061 38.86570, 0.05452 38.86324, ...  
14  POLYGON ((-1.48848 39.41741, -1.48885 39.40166...  
15  POLYGON ((-5.04694 38.72909, -5.05064 38.72628...  
16  MULTIPOLYGON (((-4.14975 39.57363, -4.13585 39...  
17  MULTIPOLYGON (((-0.87408 39.84835, -0.86360 39...  
18  POLYGON ((-1.78752 43.39112, -1.79797 43.38131...  
19  POLYGON ((-4.30627 40.22168, -4.295

In [ ]:
# Drop some columns
spanish_provinces = spanish_provinces.drop(['population', 'time_light_sun'], axis=1)
spanish_provinces

rotulo                                           geometry
0              Melilla  POLYGON ((-2.93446 35.27442, -2.92929 35.27324...
1                Ceuta  POLYGON ((-5.33273 35.89270, -5.31086 35.88727...
2                Cádiz  POLYGON ((-5.14234 37.00387, -5.11510 36.98897...
3               Málaga  POLYGON ((-4.32759 37.18435, -4.32333 37.18075...
4              Almería  POLYGON ((-2.18494 37.90366, -2.17201 37.88872...
5              Granada  POLYGON ((-2.34163 38.02603, -2.32309 38.01401...
6              Sevilla  POLYGON ((-5.72320 38.19244, -5.70785 38.18830...
7               Huelva  POLYGON ((-6.18021 37.94103, -6.14518 37.92050...
8                 Jaén  POLYGON ((-2.55129 38.08412, -2.57299 38.07785...
9              Córdoba  POLYGON ((-5.00801 38.71523, -4.87696 38.68610...
10  Castelló/Castellón  POLYGON ((-0.07991 40.73335, -0.06476 40.72742...
11              Murcia  POLYGON ((-1.11971 38.73774, -1.11563 38.71141...
12       Illes Balears  MULTIPOLYGON (((3.19286 39.93983, 3.15416 39.9...
13    Alacant/Alicante  POLYGON ((0.01061 38.86570, 0.05452 38.86324, ...
14            Albacete  POLYGON ((-1.48848 39.41741, -1.48885 39.40166...
15             Badajoz  POLYGON ((-5.04694 38.72909, -5.05064 38.72628...
16         Ciudad Real  MULTIPOLYGON (((-4.14975 39.57363, -4.13585 39...
17   València/Valencia  MULTIPOLYGON (((-0.87408 39.84835, -0.86360 39...
18            Gipuzkoa  POLYGON ((-1.78752 43.39112, -1.79797 43.38131...
19              Toledo  POLYGON ((-4.30627 40.22168, -4.29592 40.21909...
20             Cáceres  POLYGON ((-6.19123 40.47214, -6.16921 40.45812...
21              Madrid  POLYGON ((-3.39465 41.00025, -3.44917 40.87717...
22              Cuenca  POLYGON ((-1.80640 40.39824, -1.79013 40.39754...
23               Ávila  POLYGON ((-4.72211 41.14930, -4.71996 41.13487...
24         Guadalajara  POLYGON ((-2.86692 41.27446, -2.82180 41.25836...
25           Salamanca  POLYGON ((-6.47753 41.28756, -6.44532 41.28384...
26              Teruel  POLYGON ((-0.49303 41.35112, -0.47478 41.32408...
27             Segovia  POLYGON ((-3.88687 41.53643, -3.87516 41.53472...
28           Tarragona  POLYGON ((1.64513 41.19565, 1.52652 41.18085, ...
29               Soria  POLYGON ((-2.33220 42.14601, -2.30788 42.14023...
30              Zamora  POLYGON ((-5.52088 42.09711, -5.48480 41.98454...
31          Valladolid  MULTIPOLYGON (((-5.06819 42.26116, -5.05481 42...
32           Barcelona  POLYGON ((2.77827 41.64885, 2.75074 41.64420, ...
33              Girona  MULTIPOLYGON (((1.99049 42.35948, 2.01509 42.3...
34            La Rioja  POLYGON ((-2.85815 42.63818, -2.84263 42.62911...
35             Ourense  POLYGON ((-7.07701 42.50815, -7.02864 42.49783...
36             Navarra  MULTIPOLYGON (((-1.60809 43.25226, -1.57485 43...
37            Zaragoza  POLYGON ((-0.90116 42.74271, -0.90279 42.70217...
38              Lleida  POLYGON ((0.90621 42.80120, 0.93291 42.79094, ...
39              Huesca  POLYGON ((-0.74593 42.92189, -0.72761 42.91925...
40          Pontevedra  MULTIPOLYGON (((-8.13900 42.84307, -8.04587 42...
41              Burgos  MULTIPOLYGON (((-3.22356 43.17222, -3.18701 43...
42                León  POLYGON ((-5.42159 42.03065, -5.42478 42.03564...
43           Cantabria  POLYGON ((-3.15337 43.35325, -3.15097 43.34883...
44            Asturias  POLYGON ((-4.51231 43.39323, -4.52299 43.36854...
45                Lugo  POLYGON ((-7.03184 43.54449, -7.04842 43.50104...
46            A Coruña  POLYGON ((-7.69975 43.73515, -7.69602 43.71373...
47             Bizkaia  POLYGON ((-2.74784 43.44460, -2.73800 43.42922...
48         Araba/Álava  POLYGON ((-2.96449 43.17868, -2.93892 43.14526...
49            Palencia  MULTIPOLYGON (((-4.38920 43.02569, -4.37439 43...

In [ ]:
# Get geometry of Santa Cruz de Tenerife and Las Palmas
canary_islands = gpd.read_file(data_dir_geom + 'recintos_provinciales_inspire_canarias_wgs84.shp')
canary_islands

INSPIREID COUNTRY  \
0  ES.IGN.BDDAE.34053500000      ES   
1  ES.IGN.BDDAE.34053800000      ES   

                                              NATLEV NATLEVNAME      NATCODE  \
0  https://inspire.ec.europa.eu/codelist/Administ...  Provincia  34053500000   
1  https://inspire.ec.europa.eu/codelist/Administ...  Provincia  34053800000   

                 NAMEUNIT CODNUT1 CODNUT2 CODNUT3  \
0              Las Palmas     ES7    ES70    None   
1  Santa Cruz de Tenerife     ES7    ES70    None   

                                            geometry  
0  MULTIPOLYGON (((-15.69749 27.77109, -15.69750 ...  
1  MULTIPOLYGON (((-18.00161 27.64707, -18.00158 ...

In [ ]:
# Define new rows to add
row1 = pd.DataFrame([[canary_islands.loc[0, 'NAMEUNIT'], canary_islands.loc[0, 'geometry']]], columns=spanish_provinces.columns)
row2 = pd.DataFrame([[canary_islands.loc[1, 'NAMEUNIT'], canary_islands.loc[1, 'geometry']]], columns=spanish_provinces.columns)

In [ ]:
# Add new rows with geometry
spanish_provinces = pd.concat([spanish_provinces, row1, row2], ignore_index=True)
spanish_provinces

rotulo                                           geometry
0                  Melilla  POLYGON ((-2.93446 35.27442, -2.92929 35.27324...
1                    Ceuta  POLYGON ((-5.33273 35.89270, -5.31086 35.88727...
2                    Cádiz  POLYGON ((-5.14234 37.00387, -5.11510 36.98897...
3                   Málaga  POLYGON ((-4.32759 37.18435, -4.32333 37.18075...
4                  Almería  POLYGON ((-2.18494 37.90366, -2.17201 37.88872...
5                  Granada  POLYGON ((-2.34163 38.02603, -2.32309 38.01401...
6                  Sevilla  POLYGON ((-5.72320 38.19244, -5.70785 38.18830...
7                   Huelva  POLYGON ((-6.18021 37.94103, -6.14518 37.92050...
8                     Jaén  POLYGON ((-2.55129 38.08412, -2.57299 38.07785...
9                  Córdoba  POLYGON ((-5.00801 38.71523, -4.87696 38.68610...
10      Castelló/Castellón  POLYGON ((-0.07991 40.73335, -0.06476 40.72742...
11                  Murcia  POLYGON ((-1.11971 38.73774, -1.11563 38.71141...
12           Illes Balears  MULTIPOLYGON (((3.19286 39.93983, 3.15416 39.9...
13        Alacant/Alicante  POLYGON ((0.01061 38.86570, 0.05452 38.86324, ...
14                Albacete  POLYGON ((-1.48848 39.41741, -1.48885 39.40166...
15                 Badajoz  POLYGON ((-5.04694 38.72909, -5.05064 38.72628...
16             Ciudad Real  MULTIPOLYGON (((-4.14975 39.57363, -4.13585 39...
17       València/Valencia  MULTIPOLYGON (((-0.87408 39.84835, -0.86360 39...
18                Gipuzkoa  POLYGON ((-1.78752 43.39112, -1.79797 43.38131...
19                  Toledo  POLYGON ((-4.30627 40.22168, -4.29592 40.21909...
20                 Cáceres  POLYGON ((-6.19123 40.47214, -6.16921 40.45812...
21                  Madrid  POLYGON ((-3.39465 41.00025, -3.44917 40.87717...
22                  Cuenca  POLYGON ((-1.80640 40.39824, -1.79013 40.39754...
23                   Ávila  POLYGON ((-4.72211 41.14930, -4.71996 41.13487...
24             Guadalajara  POLYGON ((-2.86692 41.27446, -2.82180 41.25836...
25               Salamanca  POLYGON ((-6.47753 41.28756, -6.44532 41.28384...
26                  Teruel  POLYGON ((-0.49303 41.35112, -0.47478 41.32408...
27                 Segovia  POLYGON ((-3.88687 41.53643, -3.87516 41.53472...
28               Tarragona  POLYGON ((1.64513 41.19565, 1.52652 41.18085, ...
29                   Soria  POLYGON ((-2.33220 42.14601, -2.30788 42.14023...
30                  Zamora  POLYGON ((-5.52088 42.09711, -5.48480 41.98454...
31              Valladolid  MULTIPOLYGON (((-5.06819 42.26116, -5.05481 42...
32               Barcelona  POLYGON ((2.77827 41.64885, 2.75074 41.64420, ...
33                  Girona  MULTIPOLYGON (((1.99049 42.35948, 2.01509 42.3...
34                La Rioja  POLYGON ((-2.85815 42.63818, -2.84263 42.62911...
35                 Ourense  POLYGON ((-7.07701 42.50815, -7.02864 42.49783...
36                 Navarra  MULTIPOLYGON (((-1.60809 43.25226, -1.57485 43...
37                Zaragoza  POLYGON ((-0.90116 42.74271, -0.90279 42.70217...
38                  Lleida  POLYGON ((0.90621 42.80120, 0.93291 42.79094, ...
39                  Huesca  POLYGON ((-0.74593 42.92189, -0.72761 42.91925...
40              Pontevedra  MULTIPOLYGON (((-8.13900 42.84307, -8.04587 42...
41                  Burgos  MULTIPOLYGON (((-3.22356 43.17222, -3.18701 43...
42                    León  POLYGON ((-5.42159 42.03065, -5.42478 42.03564...
43               Cantabria  POLYGON ((-3.15337 43.35325, -3.15097 43.34883...
44                Asturias  POLYGON ((-4.51231 43.39323, -4.52299 43.36854...
45                    Lugo  POLYGON ((-7.03184 43.54449, -7.04842 43.50104...
46                A Coruña  POLYGON ((-7.69975 43.73515, -7.69602 43.71373...
47                 Bizkaia  POLYGON ((-2.74784 43.44460, -2.73800 43.42922...
48             Araba/Álava  POLYGON ((-2.96449 43.17868, -2.93892 43.14526...
49                Palencia  MULTIPOLYGON (((-4.38920 43.02569, -4.37439 43...
50              Las Palmas  MULTIPOLYGON (

In [ ]:
poly_dict = {}

for i in spanish_provinces.index:

  poly_dict[spanish_provinces['rotulo'][i]] = spanish_provinces['geometry'][i]

polygons = gpd.GeoSeries(poly_dict)
polygons

Melilla                   POLYGON ((-2.93446 35.27442, -2.92929 35.27324...
Ceuta                     POLYGON ((-5.33273 35.89270, -5.31086 35.88727...
Cádiz                     POLYGON ((-5.14234 37.00387, -5.11510 36.98897...
Málaga                    POLYGON ((-4.32759 37.18435, -4.32333 37.18075...
Almería                   POLYGON ((-2.18494 37.90366, -2.17201 37.88872...
Granada                   POLYGON ((-2.34163 38.02603, -2.32309 38.01401...
Sevilla                   POLYGON ((-5.72320 38.19244, -5.70785 38.18830...
Huelva                    POLYGON ((-6.18021 37.94103, -6.14518 37.92050...
Jaén                      POLYGON ((-2.55129 38.08412, -2.57299 38.07785...
Córdoba                   POLYGON ((-5.00801 38.71523, -4.87696 38.68610...
Castelló/Castellón        POLYGON ((-0.07991 40.73335, -0.06476 40.72742...
Murcia                    POLYGON ((-1.11971 38.73774, -1.11563 38.71141...
Illes Balears             MULTIPOLYGON (((3.19286 39.93983, 3.15416 39.9...
Alacant/Alic

In [ ]:
points_tourism = gpd.GeoDataFrame(index=tourism_cleaned['id'], crs='epsg:4326', geometry= list(tourism_cleaned['geometry']))

In [ ]:
# Check in which province each establishment of the dataset is located
estab_tourism = points_tourism.assign(**{key: points_tourism.within(geom) for key, geom in polygons.items()})
estab_tourism

geometry  Melilla  Ceuta  Cádiz  Málaga  \
id                                                                      
25913327     POINT (-3.78822 40.39839)    False  False  False   False   
26860857     POINT (-3.68898 40.31730)    False  False  False   False   
26860858     POINT (-0.16352 38.98598)    False  False  False   False   
26860866     POINT (-3.90750 40.38111)    False  False  False   False   
26860889      POINT (3.20000 41.94028)    False  False  False   False   
...                                ...      ...    ...    ...     ...   
1228903226  POINT (-15.70544 27.78727)    False  False  False   False   
1248811338  POINT (-16.75569 28.31900)    False  False  False   False   
1250033538  POINT (-16.50087 28.35548)    False  False  False   False   
1250730576  POINT (-17.97523 27.80983)    False  False  False   False   
1290130532  POINT (-18.00115 27.77478)    False  False  False   False   

            Almería  Granada  Sevilla  Huelva   Jaén  ...   León  Cantabria  \
id                                                    ...                     
25913327      False    False    False   False  False  ...  False      False   
26860857      False    False    False   False  False  ...  False      False   
26860858      False    False    False   False  False  ...  False      False   
26860866      False    False    False   False  False  ...  False      False   
26860889      False    False    False   False  False  ...  False      False   
...             ...      ...      ...     ...    ...  ...    ...        ...   
1228903226    False    False    False   False  False  ...  False      False   
1248811338    False    False    False   False  False  ...  False      False   
1250033538    False    False    False   False  False  ...  False      False   
1250730576    False    False    False   False  False  ...  False      False   
1290130532    False    False    False   False  False  ...  False      False   

            Asturias   Lugo  A Coruña  Bizkaia  Araba/Álava  Palencia  \
id                                                                      
25913327       False  False     False    False        False     False   
26860857       False  False     False    False        False     False   
26860858       False  False     False    False        False     False   
26860866       False  False     False    False        False     False   
26860889       False  False     False    False        False     False   
...              ...    ...       ...      ...          ...       ...   
1228903226     False  False     False    False        False     False   
1248811338     False  False     False    False        False     False   
1250033538     False  False     False    False        False     False   
1250730576     False  False     False    False        False     False   
1290130532     False  False     False    False        False     False   

            Las Palmas  Santa Cruz de Tenerife  
id                                              
25913327         False                   False  
26860857         False                   False  
26860858         False                   False  
26860866         False                   False  
26860889         False                   False  
...                ...                     ...  
1228903226        True                   False  
1248811338       False                    True  
1250033538       False                    True  
1250730576       False                    True  
1290130532       False                    True  

[21453 rows x 53 columns]

In [ ]:
estab_tourism_number = pd.DataFrame()

for column in estab_tourism:

  if column != 'geometry':

    number = np.count_nonzero(estab_tourism[column] == True)

    new_row = {'province':column, 'establishments':number}

    estab_tourism_number = pd.concat([estab_tourism_number, pd.DataFrame([new_row])], ignore_index=True)

estab_tourism_number

province  establishments
0                  Melilla              10
1                    Ceuta               2
2                    Cádiz             387
3                   Málaga             874
4                  Almería             218
5                  Granada             397
6                  Sevilla             324
7                   Huelva             135
8                     Jaén             304
9                  Córdoba             174
10      Castelló/Castellón             197
11                  Murcia             217
12           Illes Balears            1494
13        Alacant/Alicante             501
14                Albacete             135
15                 Badajoz             205
16             Ciudad Real             124
17       València/Valencia             383
18                Gipuzkoa             261
19                  Toledo             178
20                 Cáceres             297
21                  Madrid            1066
22                  Cuenca             127
23                   Ávila             353
24             Guadalajara             155
25               Salamanca             430
26                  Teruel             193
27                 Segovia             162
28               Tarragona             271
29                   Soria             178
30                  Zamora             166
31              Valladolid             132
32               Barcelona            1224
33                  Girona             763
34                La Rioja             149
35                 Ourense             136
36                 Navarra             547
37                Zaragoza             207
38                  Lleida             531
39                  Huesca             632
40              Pontevedra             332
41                  Burgos             581
42                    León             433
43               Cantabria             559
44                Asturias             704
45                    Lugo             341
46                A Coruña             623
47                 Bizkaia             303
48             Araba/Álava             120
49                Palencia             167
50              Las Palmas            1143
51  Santa Cruz de Tenerife             831

As can be seen, we have already calculated the number of establishments in each province in 2019.

In [ ]:
# Save the results in an Excel file
estab_tourism_number.to_excel(data_dir_new + '2019_estab_by_province.xlsx')

Once we have saved this data in a file for later use, we are going to visualize this numerical data in a graph to check more easily which provinces had the highest tourism capacity in 2019.

In [ ]:
fig = px.bar(estab_tourism_number, x='province', y='establishments')
fig.show()

As can be seen, the provinces with the greatest number of tourist establishments are Illes Balears, Barcelona, Las Palmas, Madrid, Málaga, Santa Cruz de Tenerife and Girona.